api_app/serializers.py

```diff
@@ -592,3 +597,77 @@ class EnvelopeSerializer(rfs.Serializer):
         results = serializer_class(many=many)
 
     return EnvelopeSerializer
+
+
+class CustomConfigSerializer(rfs.ModelSerializer):
+    def validate_value(self, value):
+        try:
+            json.loads(value)
+        except json.JSONDecodeError:
+            raise ValidationError("Value is not JSON-compliant.")
+        return value
+
+    def validate(self, attrs):
+        super().validate(attrs)
+
+        # check if owner is admin of organization
+        if attrs.get("organization", None):
+            # check if the user is owner of the organization
+            membership = Membership.objects.filter(
+                user=attrs.get("owner"),
+                organization=attrs.get("organization"),
+                is_owner=True,
+            )
+            if not membership.exists():
+                raise ValidationError("User is not owner of the organization.")
+
+        config = None
+        category = None
+        if attrs["type"] == CustomConfig.Type.ANALYZER:
+            config = AnalyzerConfig
+            category = "Analyzer"
+        elif attrs["type"] == CustomConfig.Type.CONNECTOR:
+            config = ConnectorConfig
+            category = "Connector"
+
+        if attrs["name"] not in config.all():
+            raise ValidationError(f"{category} {attrs['name']} does not exist.")
+
+        if attrs["attribute"] not in config.all()[attrs["name"]].params:
+            raise ValidationError(
+                f"{category} {attrs['name']} does not "
+                f"have attribute {attrs['attribute']}."
+            )
+
+        # Check if the type of value is valid for the attribute.
+        # NOTE: json.loads is used as it allows us to store all
+        # valid value types into a StringField
+        if not isinstance(
+            json.loads(attrs["value"]),
+            type(config.all()[attrs["name"]].params[attrs["attribute"]].value),
+        ):
+            expected_type = type(
+                config.all()[attrs["name"]].params[attrs["attribute"]].value
+            )
+            raise ValidationError(
+                f"{category} {attrs['name']} attribute "
+                f"{attrs['attribute']} has wrong type "
+                f"{type(json.loads(attrs['value'])).__name__}. Expected: "
+                f"{expected_type.__name__}."
+            )
+
+        if self.instance is None:
+            try:
+                params = attrs.copy()
+                params.pop("value")
+                print("params:", params)
+                self.instance = CustomConfig.objects.get(**params)
+                logger.info(f"CustomConfig {self.instance} found. Will be updated.")
+            except CustomConfig.DoesNotExist:
+                # No matching CustomConfig was found. A new entry will be created.
+                pass
```

Comment:
Ok this is to update an already existing entry with the new value, am I right? Is this necessary?

Reply:
This eliminates the need for front end to send PATCH requests. If a user has a lot of configs, they don't first have to find if a matching one exists, they can directly overwrite it.

Reply:
What you said is clever and makes sense. Nevertheless this is a bit hacky, you know? We are basically circumventing the REST API rules to avoid to an additional HTTP request. Even if I appreciate the effort, I still prefer to follow the framework and the normal rules. It makes sense that if the config never existed then we use a PUT (`update`) and when it already exists we use a PATCH (`partial_update`). In that way we are complaint

Reply:
If the config never existed, won't it be a `POST`?

Reply:
Also, so an error if a POST has a matching entry?

Reply:
yeah you are right, I meant that, God I am melted 🤣  POST `create` 

Then yeah I think yes because after the first time we set it, we only want to update the value, nothing else, am I right?

Reply:
`DELETE` too, but that's anyways handled